Fix issues with TS compiler flags (isolatedDeclarations + erasableSyntaxOnly)

blutorange opened the pull request:
This is a great project, I use it for a smaller DSL where it's still too annoying to code the parser manually.

I had a few issue with newer TS compiler flags, which this PR fixes. In particular:

* [--isolatedDeclarations](https://www.typescriptlang.org/tsconfig/#isolatedDeclarations)
  * Requires explicit return types. The intention is to make it easier for third-party tools to generate declaration files, and to improve performance.
  * There were two method that were missing a `void` return type, which I added.
* [--erasableSyntaxOnly](https://www.typescriptlang.org/tsconfig/#erasableSyntaxOnly)
  * TypeScript's current philosophy is not to add any runtime features. Enum are an old construct that violate this design constraint. The new `erasableSyntaxOnly` explicitly forbids such constructs. `enum` is such a construct.
  * When running a TypeScript file with the current node version, it results in the error `SyntaxError [ERR_UNSUPPORTED_TYPESCRIPT_SYNTAX]: TypeScript enum is not supported in strip-only mode`
  * TypeScript now supports union types, which are a good alternative to `enum`. 
  * I added a `--union-enums` flag (default: `false`) that causes the generator to use unions instead of enums.

awa-xima committed:
Add explciit return type to generated functions

> Method must have an explicit return type annotation with --isolatedDeclarations.ts(9008)

awa-xima committed:
Add flag --union-enums to use unions instead of TS enums

- SyntaxError [ERR_UNSUPPORTED_TYPESCRIPT_SYNTAX]: TypeScript enum is not supported in strip-only mode
- This syntax is not allowed when 'erasableSyntaxOnly' is enabled.ts(1294)

awa-xima committed:
Update package-lock.json

awa-xima committed:
Updated generated test parsers

awa-xima committed:
Fix EofType in template.ts

blutorange committed:
Add note regarding --union-enums flag to readme

EoinDavey commented:
Hey, thanks for the work and the nice feedback! 
I have a question about the union-enums and `erasableSyntaxOnly`, I haven't been working on much typescript recently so this is new to me. I'm wondering if this is somethign I should enable by default? In particular, would this break existing projects if I enable this globally?

If it's a "safe change" to apply in general, then I wonder if we need the CLI flag `--union-enums` at all

blutorange commented:
>  I have a question about the union-enums and erasableSyntaxOnly. I haven't been working on much typescript recently so this is new to me. I'm wondering if this is something I should enable by default? In particular, would this break existing projects if I enable this globally?

I think `--union-enums` should __not be enabled by default in tspeg__, as it change the generator output. `ASTKinds`  is not an `enum` anymore, but a union of string/number constants + a union type. It behaves pretty similar, but there are subtle differences. So take the test as an example:

[TypeScript playground](https://www.typescriptlang.org/play/?#code/PTAEF5K6ZgoEFKgM4BcBOBLAdgc1AFMcBXAWxSXATFjrrhtAHUs0ALAexLVAFo+JHFk44+xcijiEAHgAdOGXhLKgAggGUAKgGlcAExQBRUqoDeoDQFUAsgBpLtgPoASAAwOAkgDktoAL4A3IyIrBz8gsKi4qZSsgpKoADGoujq2no4hlZROBCgFtY2AFygAERFZQ5Frm6lFc7uVaA+WvWtZQGgAIaUKTjowfGKvGgAnnKE6boGKDkieeCg5ZWgAD4rjW6dG2UdwXD9ad0AjKWaM1nGpvkXmYYm5AB0rYGgiADuigDWUke8ACMztN7nNcrcMrN5qIXr43iAvhhfiEwN5OLwUCQMFMPuwxj0cGMAZx9PiviQADb6UA4QiEaloTigfRMjhYFAOAE8UAcQj4lKU-SHVK8boAJnOkKujzITyKEMuD1Mcts8LAiOR-1AALFwLuUNyKpsCtB0JwRrVyU4ZDkWApU0IGAwihRVHo7uoiCgoG85EdWCSRFiVCYHrDrrCXG5AiECxikmk8hGQfIINmMt95kcxqWHmztXyJy8vnyYoCB1CbHYEVj0RUcSTiS1+quZsz+UKtlKeZq7lKRZavlKZf8PT6IqGjdGEymLeyuXbufWoBOy7FBy1OFKpgBjvyAFZgogcEyUja7Q6nS6tUk9VKleRFzS3kwz7b7UQrxhhQNePo74qYILE+OAvkAA)

```ts
// ====================
// === string enums ===
// ====================

// Without --union-enums
export enum ASTKindsEnum { SUM, SUM_$0, INT };

// With --union-enums
export const ASTKindsUnion = { SUM: "SUM", SUM_$0: "SUM_$0", INT: "INT" } as const;
export type ASTKindsUnion =  "SUM" | "SUM_$0" | "INT";

const a1: ASTKindsEnum = ASTKindsEnum.INT; // works
const b1: ASTKindsUnion = ASTKindsUnion.INT; //works

// Not sure why anybody would need to do this, but they could
// works
const a2: ASTKindsEnum.SUM = ASTKindsEnum.SUM;
// compile error
const b21: ASTKindsUnion.SUM = ASTKindsUnion.SUM;

// =====================
// === Numeric enums ===
// =====================

// Without --union-enums
export enum ASTKindsEnumNum { SUM = 0, SUM_$0 = 1, INT = 2 };

// With --union-enums
export const ASTKindsUnionNum = { SUM: 0, SUM_$0: 1, INT: 2 } as const;
export type ASTKindsUnionNum = 0 | 1 | 2;

// Again not sure why anybody would need to do this, but they could
const n: number = 5;
// no compile error
const c1: ASTKindsEnumNum = n; 
// compile error
const d1: ASTKindsUnionNum = n; 
```

EoinDavey commented on package-lock.json:
This version number shouldn't be bumped yet

EoinDavey commented on package-lock.json:
This version number shouldn't be bumped yet

blutorange replied:
`npm install` updated the version in the `package-lock.json`, as the `package.json` [has 3.3.2 already](https://github.com/EoinDavey/tsPEG/blob/master/package.json#L3). I can change it back in the package-lock, but out of curiosity, is there a reason for this mismatch?

EoinDavey replied:
That's interesting, I have no explanation for it tbh. I'll update the package-lock.json to match. Thanks!

EoinDavey commented on src/cli.ts:
```diff
@@ -36,6 +36,11 @@ yargs.command("$0 <grammar> [output_file]", "Build parser from grammar",
                 default: false,
                 desc: "Use numeric enums for AST kinds",
             },
+            "union-enums": {
```
I think this flag is too low-level, I think it should be called "erasable-syntax" so that it represents why you might use it, rather than it's effect

blutorange replied:
Just for the record, I personally also prefer string constant unions over enums in general, it's also e.g. what the [acorn parser uses](https://github.com/acornjs/acorn/blob/master/acorn/src/acorn.d.ts). So I can just use a string without having to import some enum, e.g.

```ts
function processNode(node) {
  switch (node.kind) {
     case "sum":
       return node.left. + node.right;
     case "product":
       return node.left. * node.right;
}
```

That it also works with `--erasableSyntaxOnly` is a bonus -- I would probably enable the flag anyways.  But perhaps that's just me, I'll rename the option for now.

EoinDavey commented on src/cli.ts:
```diff
@@ -36,6 +36,11 @@ yargs.command("$0 <grammar> [output_file]", "Build parser from grammar",
+            "union-enums": {
+                type: "boolean",
+                default: false,
+                desc: "Use unions instead of enums for AST kinds",
```
Can you update the description also. Maybe include a link to https://www.typescriptlang.org/tsconfig/#erasableSyntaxOnly

EoinDavey commented on src/gen.ts:
```diff
@@ -74,10 +74,12 @@ export class Generator {
     private checkers: Checker[] = [];
     private header: string | null;
     private boundedRecRules: Set<string>;
+    private astKindsByName: Map<string, string> = new Map();
```
I think it's not ideal that this is a class-level field, but shouldn't be accessed until `writeKinds` is called. Could we move the computation of `astKinds` from the `writeKinds` method to the constructor so that this map can also be filled out by the constructor?

EoinDavey commented on src/gen.ts:
```diff
@@ -429,14 +448,22 @@ export class Generator {
+
+    private astKindsType(name: string): string {
+        if (this.unionEnums) {
+            return this.astKindsByName.get(name) ?? "never"; 
```
I'm assuming that we don't expect the `?? "never"` case to happen? It's just satisfying the type checker?

blutorange replied:
Yes it shouldn't happen, the map should contain a value for the name (or it's a bug). But I like a logical default better than just a non-null assertion `return this.astKindsByName.get(name)!`

EoinDavey commented:
Thanks for explaining why it's not a safe change, makes sense to me.
I just reviewed the changes, overall looks great! Just requested some small changes. Thanks for the work!

blutorange committed:
Minor adjustments

* Downgrade version in package-lock
* Improve description for union-enums
* Make astKindsByName an immutable class field

blutorange committed:
Rename --union-enums to --erasable-syntax

This describes its purpose, not its effect.

But keep the original name in the internal implementation,
as we want to know what it does in the implementation.

awa-xima committed:
Finish renaming unionEnums to erasableSyntax

If this is what we call the flag, then its effect must be
to ensure generated code compiles under the --erasable-syntax
flag. That includes e.g. not using enum, but is not limited to
this change (potentially).

blutorange commented:
Anyways, I made the changes, let me know if you'd like to change anything else.

PS: With the new features from node and TypeScript, in my project I can now use the native Node.JS test runner from the module `node:test` and run TypeScript test files directly, without transpilation or any other external test library.
